test/integration/parse_tree_visitor_test.rb

```diff
@@ -0,0 +1,231 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class ParseTreeVisitorTest < Minitest::Test
+  include Liquid
+
+  def traversal(template)
+    ParseTreeVisitor
+      .for(Template.parse(template).root)
+      .add_callback_for(VariableLookup, &:name)
+      .visit.flatten.compact
```

Comment:
Why the `visit.flatten.compact` here? by making this a part of the test, I might accidentally break the API that you wanted without your tests covering it. For example I could change the code to always add a `nil` to the return value of every `visit` and that would most likely break consumers of this new API, but none of your tests would cover that. Same for the `flatten` (I could add some garbage array nesting and your tests wouldn't catch it).

I think you should add at least one or two tests or so to explicitly assert that the array structure is what you wanted. If you *always* want this to be flat and compact then maybe `#visit` itself should do that (rather than the caller)?